**Spring 3:** Alexis La Cruz

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# se importan las bases de datos
u2 = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u2.base')
ua = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/ua.base')
ub = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/ub.base')
u3 = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u3.base')
u4 = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u4.base')
u5 = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u5.base')
udata = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u.data')
u2t = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u2.test',)
uat = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/ua.test')
ubt = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/ub.test')
u3t = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u3.test')
u4t = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u4.test')
u5t = pd.read_table('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u5.test')
genre = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u.genre', sep='|', encoding='latin-1')
occupation = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u.occupation', sep='|', encoding='latin-1')
user = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u.user', sep='|', encoding='latin-1')
item = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science/Springs/Tercer Spring/archivos/u.item', sep='|', encoding='latin-1')

In [3]:
print(u2.shape)
print(ua.shape)
print(ub.shape)
print(u3.shape)
print(u4.shape)
print(u5.shape)

(79999, 4)
(90569, 4)
(90569, 4)
(79999, 4)
(79999, 4)
(79999, 4)


Las bases de datos del train tienen 4 columnas. 4 de las bases de datos tienen 79999 filas y 2 90569 filas.

In [4]:
print(u2t.shape)
print(uat.shape)
print(ubt.shape)
print(u3t.shape)
print(u4t.shape)
print(u5t.shape)

(19999, 4)
(9429, 4)
(9429, 4)
(19999, 4)
(19999, 4)
(19999, 4)


Las bases de datos del test tienen 4 columnas. 4 de las bases de datos tienen 19999 filas y 2 9429 filas.

In [5]:
# se unen las bases de datos del train
df_train = pd.concat([ua,ub,u2,u3,u4,u5],ignore_index=True)

In [6]:
# se unen las bases de datos para generar el dataset de validación
df_test = pd.concat([uat,ubt,u2t,u3t,u4t,u5t],ignore_index=True)

In [7]:
df_train = df_train.drop(['3','4','878542960'],axis=1)
df_test = df_test.drop(['17','3','875073198','1.1','5','874965758','889751712','876893119','878542960'],axis=1)

In [8]:
df_train.columns = ['user_id', 'item_id', 'rating', 'timestamp']
df_test.columns = ['user_id', 'item_id', 'rating', 'timestamp']

In [9]:
df_train.isnull().sum()/len(df_train)

user_id      0.000000
item_id      0.159636
rating       0.159636
timestamp    0.159636
dtype: float64

In [10]:
df_test.isnull().sum()/len(df_test)

user_id      0.000000
item_id      0.904617
rating       0.500000
timestamp    0.904617
dtype: float64

In [11]:
# se suprimen los valores nulos surgidos de la combinación de los datasets
df_train = df_train.dropna()
df_test = df_test.dropna()

In [12]:
# a continuación se genera un tercer dataset con datos aleatorios sustraídos del train y del test
df_val1 = df_train.sample(n=10000)
df_val2 = df_test.sample(n=1000)
df_val = pd.concat([df_val1,df_val2],ignore_index=True)

In [13]:
# se cambia el formato de los valores de la columna "timestamp" a fecha
from datetime import datetime
df_train["timestamp"] = df_train["timestamp"].apply(lambda x: datetime.utcfromtimestamp(x).strftime('%Y-%m-%d'))
df_val["timestamp"] = df_val["timestamp"].apply(lambda x: datetime.utcfromtimestamp(x).strftime('%Y-%m-%d'))
df_test["timestamp"] = df_test["timestamp"].apply(lambda x: datetime.utcfromtimestamp(x).strftime('%Y-%m-%d'))

In [14]:
# se procede a construir la matriz de interacciones
columnas = ["user_id", "item_id", "rating"]
interacciones = df_train[columnas]

In [15]:
interacciones = interacciones.drop_duplicates(['user_id'])

In [16]:
matriz_de_interacciones = interacciones.pivot(index="user_id", columns="item_id", values="rating")

In [17]:
matriz_de_interacciones.head()

item_id  1.0    2.0    3.0    4.0    5.0    6.0    7.0    8.0    9.0    10.0   \
user_id                                                                         
1          NaN    3.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2          4.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5          NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

item_id  ...  264.0  266.0  268.0  269.0  270.0  272.0  286.0  288.0  289.0  \
user_id  ...                                                                  
1        ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2        ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3        ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4        ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5        ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

item_id  302.0  
user_id         
1          NaN  
2          NaN  
3          NaN  
4          NaN  
5          NaN  

[5 rows x 76 columns]

In [18]:
# se reemplazan los NaN por 0
matriz_de_interacciones = matriz_de_interacciones.fillna(0)

In [19]:
matriz_de_interacciones.head()

item_id  1.0    2.0    3.0    4.0    5.0    6.0    7.0    8.0    9.0    10.0   \
user_id                                                                         
1          0.0    3.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2          4.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5          0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

item_id  ...  264.0  266.0  268.0  269.0  270.0  272.0  286.0  288.0  289.0  \
user_id  ...                                                                  
1        ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2        ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3        ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4        ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5        ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

item_id  302.0  
user_id         
1          0.0  
2          0.0  
3          0.0  
4          0.0  
5          0.0  

[5 rows x 76 columns]

In [20]:
print(matriz_de_interacciones.shape)

(943, 76)


La matriz de interacciones tiene 943 filas y 76 columnas

In [21]:
# se procede a transformar la matriz a csr matrix dado que esta es muy grande y dispersa
from scipy.sparse import csr_matrix
matriz_de_interacciones_csr = csr_matrix(matriz_de_interacciones.values)
matriz_de_interacciones_csr

<943x76 sparse matrix of type '<class 'numpy.float64'>'
	with 943 stored elements in Compressed Sparse Row format>

In [22]:
user_ids = list(matriz_de_interacciones.index)
user_dict = {}
counter = 0 
for i in user_ids:
    user_dict[i] = counter
    counter += 1

In [23]:
user_dict

{1: 0,
 2: 1,
 3: 2,
 4: 3,
 5: 4,
 6: 5,
 7: 6,
 8: 7,
 9: 8,
 10: 9,
 11: 10,
 12: 11,
 13: 12,
 14: 13,
 15: 14,
 16: 15,
 17: 16,
 18: 17,
 19: 18,
 20: 19,
 21: 20,
 22: 21,
 23: 22,
 24: 23,
 25: 24,
 26: 25,
 27: 26,
 28: 27,
 29: 28,
 30: 29,
 31: 30,
 32: 31,
 33: 32,
 34: 33,
 35: 34,
 36: 35,
 37: 36,
 38: 37,
 39: 38,
 40: 39,
 41: 40,
 42: 41,
 43: 42,
 44: 43,
 45: 44,
 46: 45,
 47: 46,
 48: 47,
 49: 48,
 50: 49,
 51: 50,
 52: 51,
 53: 52,
 54: 53,
 55: 54,
 56: 55,
 57: 56,
 58: 57,
 59: 58,
 60: 59,
 61: 60,
 62: 61,
 63: 62,
 64: 63,
 65: 64,
 66: 65,
 67: 66,
 68: 67,
 69: 68,
 70: 69,
 71: 70,
 72: 71,
 73: 72,
 74: 73,
 75: 74,
 76: 75,
 77: 76,
 78: 77,
 79: 78,
 80: 79,
 81: 80,
 82: 81,
 83: 82,
 84: 83,
 85: 84,
 86: 85,
 87: 86,
 88: 87,
 89: 88,
 90: 89,
 91: 90,
 92: 91,
 93: 92,
 94: 93,
 95: 94,
 96: 95,
 97: 96,
 98: 97,
 99: 98,
 100: 99,
 101: 100,
 102: 101,
 103: 102,
 104: 103,
 105: 104,
 106: 105,
 107: 106,
 108: 107,
 109: 108,
 110: 109,
 111: 11

In [24]:
pip install lightfm

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.4/316.4 kB 24.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for lightfm: filename=lightfm-1.17-cp39-cp39-linux_x86_64.whl size=889514 sha256=09d400ea8a196ef686015122a465b2cb086d6571e9d2bd167d4ef31e93bd15cd
  Stored in directory: /root/.cache/pip/wheels/d8/65/93/6ac8180274dc2e8f86ff326be62da1dfa55dc158fd45faba7d
Successfully built lightfm


In [25]:
# se entrena el modelo
from lightfm import LightFM
modelo = LightFM(no_components=3, random_state=100, learning_rate=0.03)

In [26]:
modelo = modelo.fit(matriz_de_interacciones_csr, epochs=20)

In [27]:
n_users, n_items = matriz_de_interacciones.shape

In [28]:
import numpy as np
user_x = user_dict[1]
n_users, n_items = matriz_de_interacciones.shape
item_ids = np.arange(n_items)

In [29]:
predicciones = modelo.predict(user_ids=user_x, item_ids = item_ids)

In [30]:
scores = pd.Series(predicciones)
scores.index = matriz_de_interacciones.columns
recomendaciones = list(pd.Series(scores.sort_values(ascending=False).index))[:20]

A continuación se muestran los id de los contenidos recomendados para el usuario 1:

In [31]:
recomendaciones

[1.0,
 7.0,
 4.0,
 9.0,
 2.0,
 245.0,
 258.0,
 8.0,
 50.0,
 242.0,
 15.0,
 14.0,
 5.0,
 11.0,
 12.0,
 243.0,
 3.0,
 22.0,
 259.0,
 10.0]

A continuación se muestran los id de los contenidos que visualizó el usuario número 1, con rating igual al último decil.

In [32]:
df_train[(df_train.user_id==1)&(df_train.rating==df_train['rating'].quantile(q=0.9)) & (df_train.item_id.isin(recomendaciones))].item_id.unique()

array([  9.,  12.,  14.,  15.,  50., 242., 258.])

El usuario número 1 ya había visto la mayoría del contenido, sin embargo, 7 de ellos tienen un rating ubicado en el percentil 90.

In [33]:
df_train[(df_train.user_id==1) & (df_train.item_id.isin(recomendaciones))].item_id.unique()

array([  2.,   3.,   4.,   5.,   7.,   8.,   9.,  10.,  11.,  12.,  14.,
        15.,  22.,  50., 242., 243., 245., 258., 259.])

En conclusión, el sistema de recomendación parecer ser no muy bueno, dado que recomendó al usuario en cuestión contenido que en su mayoría él ya había visualizado.